Sources/TSCBasic/FileSystem.swift

```diff
@@ -184,6 +184,16 @@ public protocol FileSystem: class {
     ///
     /// The method throws if the underlying stat call fails.
     func getFileInfo(_ path: AbsolutePath) throws -> FileInfo
+
+    /// Returns the path in the normalization form (or lack thereof) actually present.
+    /// Any path components without an extant match are left unaltered.
+    ///
+    /// On file systems that natively consider Unicode‐equivalent paths to be equal,
+    /// the returned value need not represent the actual scalars present.
+    /// It is sufficient if whatever scalars are returned
+    /// *will be considered equal to the real ones by the file system*
+    /// when future file system calls are made.*
+    func resolveUnicode(_ path: AbsolutePath) -> AbsolutePath
```

Comment:
Ideally, a "unicode resolved" path would be reflected in the type system as such. Have you thought about ways to present this distinction?

Otherwise, be sure to liberally assert anywhere assuming a resolved path is in fact resolved (if it comes up). I like as many checks as possible to guard against the omission of an API call, especially as code evolves over multiple contributors.

Example:
```swift
extension FileSystem {
  internal func _assertUnicodeResolved(_ path: AboslutePath) {
#if os(macOS)
    return
#else
    assert(path == self.resolveUnicode(path))
  }
}
```

which can be sprinkled about as needed. If the idea is that all `AbsolutePath`-receiving APIs on this protocol have to be resolved, then a separate (or phantom) type becomes more appealing. Another option is to have the protocol requirements be something like `uncheckedGetFileInfo(_:)` and have `getFileInfo(_:)` be in an extension that runs this assert before calling `uncheckedGetFileInfo(_:)`.

Reply:
Also, if this is about putting the path into the conformer's preferred normal form, then this might be better called `normalizeUnicode` or something, unless normalization is the only kind of resolution imaginable for file systems (it might be).

Reply:
Thanks for taking a look, @milseman. All very good questions.

> Is canonical equivalence what you need for all platforms, or do you need compatibility equivalence? If you're trying to define a new model for SPM, then I agree with canonical equivalence. If you're trying to match an existing system's (legacy) model, then you might need compatibility equivalence.

I’m not aiming at supporting old code pages. (If someone else wants to extend it for that, I leave it up to them.) The aim here is about resolving the discrepancy between path components entered in the manifest, which use Swift `String`’s Unicode equivalence, and the file paths stored by the platform, which outside of macOS do not attempt to handle Unicode equivalence. A manifest should be equally loadable on every platform. Right now you get “file not found” errors when you go back and forth between platforms.

> Tangent: what is the story for unpaired surrogates in file names, which can happen on Windows?

I have not thought about them and I have done nothing special to handle them. Whatever `String` does ought to be fine, at least for now. In contrast to what this aims to fix, unpaired surrogates would indicate that the string itself is already “broken” to begin with.

> Ideally, a "unicode resolved" path would be reflected in the type system as such. Have you thought about ways to present this distinction?

Initially I wanted to it that way. Unfortunately, I realized that is itself unsafe, because the “resolvedness” of a path can be invalidated over time as file operations occur, some of which are hard for SwiftPM to track (such as Git). Since resolution must be redone before each operation anyway, the type‐system doesn’t actually help us much.

> Otherwise, be sure to liberally assert anywhere assuming a resolved path is in fact resolved (if it comes up).

The assertion would have to do all the same work to check whether the resolution is valid. (And we cannot just record a flag in `AbsolutePath` for the same reason a separate type would not work. Calling `resolveUnicode(_:)` at the start of each `LocalFileSystem` method, effectively accomplishes the same safety otherwise afforded by `assert`.

> Another option is to have the protocol requirements be something like uncheckedGetFileInfo(_:) and have getFileInfo(_:) be in an extension that runs this assert before calling uncheckedGetFileInfo(_:).

The two private underscored methods are basically what you mean by `unchecked`. They only exist because they need a variant callable from within `resolveUnicode(_:)` that doesn’t wind up circular. Nothing else should be using them.

> Also, if this is about putting the path into the conformer's preferred normal form, then this might be better called normalizeUnicode or something, unless normalization is the only kind of resolution imaginable for file systems (it might be).

Normalization‐related stuff is the only resolution I am imagining here. However, it is actually not normalization, but the reverse. On platforms besides macOS, the file on disk may have a name that is neither NFD nor NFC, and we still need to be able to find it.